src/stream/zstd.rs

```diff
@@ -0,0 +1,198 @@
+    state: DeState,
+    output: BytesMut,
+    decoder: Decoder,
+}
+
+impl<S: Stream<Item = Result<Bytes>>> ZstdEncoder<S> {
+    /// Creates a new encoder which will read uncompressed data from the given stream and emit a
+    /// compressed stream.
+    ///
+    /// The `level` argument here can range from 1-21. A level of `0` will use zstd's default, which is `3`.
+    pub fn new(stream: S, level: i32) -> ZstdEncoder<S> {
+        ZstdEncoder {
+            inner: stream,
+            state: State::Reading,
+            output: BytesMut::new(),
+            encoder: Encoder::new(level).unwrap(),
+        }
+    }
+}
+
+impl<S: Stream<Item = Result<Bytes>>> ZstdDecoder<S> {
+    /// Creates a new decoder which will read compressed data from the given stream and emit an
+    /// uncompressed stream.
+    pub fn new(stream: S) -> ZstdDecoder<S> {
+        ZstdDecoder {
+            inner: stream,
+            state: DeState::Reading,
+            output: BytesMut::new(),
+            decoder: Decoder::new().unwrap(),
+        }
+    }
+}
+
+impl<S: Stream<Item = Result<Bytes>>> Stream for ZstdEncoder<S> {
+    type Item = Result<Bytes>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
+        let mut this = self.project();
+
+        fn compress(
+            encoder: &mut Encoder,
+            input: &mut Bytes,
+            output: &mut BytesMut,
+        ) -> Result<Bytes> {
+            const OUTPUT_BUFFER_SIZE: usize = 8_000;
+
+            if output.len() < OUTPUT_BUFFER_SIZE {
+                output.resize(OUTPUT_BUFFER_SIZE, 0);
+            }
+
+            let status = encoder.run_on_buffers(input, output)?;
+            input.advance(status.bytes_read);
+            Ok(output.split_to(status.bytes_written).freeze())
+        }
+
+        #[allow(clippy::never_loop)] // https://github.com/rust-lang/rust-clippy/issues/4058
+        loop {
+            break match mem::replace(this.state, State::Invalid) {
+                State::Reading => {
+                    *this.state = State::Reading;
+                    *this.state = match ready!(this.inner.as_mut().poll_next(cx)) {
+                        Some(chunk) => State::Writing(chunk?),
+                        None => State::Flushing,
+                    };
+                    continue;
+                }
+                State::Writing(mut input) => {
+                    if input.is_empty() {
+                        *this.state = State::Reading;
+                        continue;
+                    }
+
+                    let chunk = compress(&mut this.encoder, &mut input, &mut this.output)?;
+
+                    *this.state = State::Writing(input);
+
+                    Poll::Ready(Some(Ok(chunk)))
+                }
+                State::Flushing => {
+                    let mut output = zstd_safe::OutBuffer::around(this.output);
+
+                    let bytes_left = this.encoder.flush(&mut output).unwrap();
```

Comment:
Can `finish` be called here instead of having to first `flush` then `finish`? The Rust libs docs don't really seem to say, but the C docs have

> Calling ZSTD_compressStream2() with ZSTD_e_end instructs to finish a frame. It will perform a flush and write frame epilogue.

which is what I assume the Rust bindings will be calling into.

You might also need to resize the `output` in this state, if there's over 8k of data to flush for some reason then you'll consume the entire output buffer on the first poll and have no space available the next time.

Reply:
So interesting story here, but I actually thought it would do the same -- I was repeatedly calling finish until I realized it didn't have the right behavior, because as I kept calling it, the bytes left in the internal buffer remained the same (32009).

Curiously enough, `flush` has the right behavior, but doesn't finish the final internal frame (as the Zstd decoder reaches unexpected EOF and returns the proper error without [line 134](https://github.com/rustasync/async-compression-rs/pull/13/files#diff-c3c3f8cc95ca9e4ee4eb756858c8a398R134)), hence the final call to finish. Actually, the boolean included doesn't even really matter in the code, it seems:

```rust
    fn finish(
        &mut self,
        output: &mut OutBuffer,
        _finished_frame: bool,
    ) -> io::Result<usize> {
        zstd_safe::end_stream(&mut self.context, output).map_err(map_error_code)
    }
```

Understandably it does look a bit strange, but when you debug the code as follows:

```rust
State::Flushing => {
    dbg!(&this.output.len()); // note the dbg here
    let mut output = zstd_safe::OutBuffer::around(this.output);

    let bytes_left = this.encoder.flush(&mut output).unwrap();
    dbg!(&output.pos, &bytes_left); // and here
    *this.state = if bytes_left == 0 {
        let _ = this.encoder.finish(&mut output, true);
        State::Done
    } else {
        State::Flushing
    };
    Poll::Ready(Some(Ok(output.as_slice().into())))
}
```
This is what the tests spit out with `cargo test zstd_stream_compress -- --nocapture`

```
running 2 tests
[src/stream/zstd.rs:130] &this.output.len() = 8000
[src/stream/zstd.rs:134] &output.pos = 15
[src/stream/zstd.rs:134] &bytes_left = 0
test zstd_stream_compress ... ok
[src/stream/zstd.rs:130] &this.output.len() = 8000
[src/stream/zstd.rs:134] &output.pos = 8000
[src/stream/zstd.rs:134] &bytes_left = 32009
[src/stream/zstd.rs:130] &this.output.len() = 8000
[src/stream/zstd.rs:134] &output.pos = 8000
[src/stream/zstd.rs:134] &bytes_left = 24009
[src/stream/zstd.rs:130] &this.output.len() = 8000
[src/stream/zstd.rs:134] &output.pos = 8000
[src/stream/zstd.rs:134] &bytes_left = 16009
[src/stream/zstd.rs:130] &this.output.len() = 8000
[src/stream/zstd.rs:134] &output.pos = 8000
[src/stream/zstd.rs:134] &bytes_left = 8009
[src/stream/zstd.rs:130] &this.output.len() = 8000
[src/stream/zstd.rs:134] &output.pos = 8000
[src/stream/zstd.rs:134] &bytes_left = 9
[src/stream/zstd.rs:130] &this.output.len() = 8000
[src/stream/zstd.rs:134] &output.pos = 9
[src/stream/zstd.rs:134] &bytes_left = 0
test zstd_stream_compress_large ... ok
```
You can note here that the "OutBuffer" that wraps around `this.output` actually just uses it as a raw byte buffer that repeatedly fills it to maximum capacity (8000) on every `flush` call when the buffer can't consume all of the bytes left in the internal buffer. Nothing actually moves in `this.output` since `output` (the outbuffer) keeps track of its' own internal position within the buffer, so it doesn't need to be resized.

Hopefully that explains my reasoning well enough, but yeah -- it was quite an interesting ride to figure out.